### PR TITLE
Add `/api/tvl/aggregate` endpoint

### PR DESCRIPTION
**Pull request by aminlatifi**
Resolve L2B-2538

### CODE REVIEW

File: packages/backend/src/api/routers/TvlRouter.ts [hunk 1]
```diff
@@ -18,6 +18,32 @@ export function createTvlRouter(
 ) {
   const router = new Router()
 
+  // endpoint that will return tvl of specific projects
+  // accepts projectIds[] as query params
+  // /api/tvl/projects + projectIds[]
+  router.get('/api/tvl/projects', async (ctx) => {
+    // get projectIds from query params
+    const projectSlugs = ctx.query.projectSlugs as string;
```
Comment (torztomasz) [inline packages/backend/src/api/routers/TvlRouter.ts:26]: You should use `withTypedContext` for validating query params

File: packages/backend/src/api/controllers/tvl/TvlController.ts [hunk 2]
```diff
@@ -32,6 +37,16 @@ type TvlResult =
       error: 'DATA_NOT_FULLY_SYNCED' | 'NO_DATA'
     }
 
+type TvlProjectResult =
```
Comment (torztomasz) [inline packages/backend/src/api/controllers/tvl/TvlController.ts:40]: This probably should be extracted to other file
`type SuccessResult<T> = { result: 'success', data: T}`
`type ErrorResult<E> = { result: 'error', data: E}`
`type Result<T, E> = SuccessResult<T> | ErrorResult<E>`

And use it in every place where we use this pattern

File: packages/backend/src/api/controllers/tvl/TvlController.ts [hunk 3]
```diff
@@ -100,6 +115,59 @@ export class TvlController {
...
+        'TVL',
+      ),
+      this.aggregatedReportRepository.getSixHourly(
+        getSixHourlyMinTimestamp(dataTimings.latestTimestamp),
+        'TVL',
+      ),
+      this.aggregatedReportRepository.getDaily('TVL'),
+    ])
+
+    const projectIdsFilterSet = new Set(projectIdsFilter)
```
Comment (torztomasz) [inline packages/backend/src/api/controllers/tvl/TvlController.ts:151]: Why do we need the Set, can't we just use the array?

**Review by torztomasz — CHANGES_REQUESTED** [review]

File: packages/backend/src/api/controllers/tvl/TvlController.ts (hunk 3 above)
Comment (aminlatifi) [reply packages/backend/src/api/controllers/tvl/TvlController.ts]: The Set has better performance, so I picked that.
I also fixed an issue, we were comparing references (projectId is not string)

File: packages/backend/src/api/routers/TvlRouter.ts [hunk 4]
```diff
@@ -18,6 +18,36 @@ export function createTvlRouter(
 ) {
   const router = new Router()
 
+  // endpoint that will return tvl of specific projects
+  // accepts projectIds[] as query params
+  // /api/tvl/projects + projectIds[]
+  router.get('/api/tvl/projects', withTypedContext(z.object({
```
Comment (antooni) [inline packages/backend/src/api/routers/TvlRouter.ts]: Maybe `/api/tvl/aggregate` ?

File: packages/backend/src/api/controllers/tvl/TvlController.ts [hunk 5]
```diff
@@ -100,6 +115,62 @@ export class TvlController {
...
+
+    const projectIdsFilterSet = new Set(
+      projectIdsFilter.map((x) => x.toString()),
+    )
+
+    const tvlApiProjectResponse = generateAggregatedApiResponse(
+      hourlyReports,
+      sixHourlyReports,
+      dailyReports,
+      this.projects
```
Comment (antooni) [inline packages/backend/src/api/controllers/tvl/TvlController.ts:159]: here you can just pass the projectIdsFilter

File: packages/backend/src/api/controllers/tvl/generateTvlApiResponse.ts [hunk 6]
```diff
@@ -38,6 +48,126 @@ export function generateTvlApiResponse(
...
+      hourlyGroupedByTimestamp[timestamp],
+    )
+    const _sixHourlyTimestamp = groupByProjectIdAndTimestamp(
+      sixHourlyGroupedByTimestamp[timestamp],
+    )
+    const _dailyTimestamp = groupByProjectIdAndTimestamp(
+      dailyGroupedByTimestamp[timestamp],
+    )
+
+    let hourlyValue: DetailedTvlApiChartPoint = [
```
Comment (antooni) [inline packages/backend/src/api/controllers/tvl/generateTvlApiResponse.ts:106]: extract this array into constant and reuse it in other places

File: packages/backend/src/api/controllers/tvl/generateTvlApiResponse.ts [hunk 7]
```diff
@@ -1,15 +1,25 @@
 import {
+  DetailedTvlApiChartPoint,
+  DetailedTvlApiCharts,
   ProjectId,
   TvlApiChart,
   TvlApiChartPoint,
   TvlApiCharts,
+  // TvlApiProjectsResponse,
```
Comment (torztomasz) [inline packages/backend/src/api/controllers/tvl/generateTvlApiResponse.ts:8]: remove comment

**Review by torztomasz — CHANGES_REQUESTED** [review]

File: packages/backend/src/api/controllers/tvl/TvlController.ts (hunk 5 above)
Comment (aminlatifi) [reply packages/backend/src/api/controllers/tvl/TvlController.ts:143]: hourly, sixHourly and daily report values are needed inside `generateAggregatedApiResponse`

File: packages/backend/package.json [hunk 8]
```diff
@@ -57,9 +57,9 @@
     "@types/basic-auth": "^1.1.3",
     "@types/deep-diff": "^1.0.2",
     "@types/koa": "^2.13.5",
+    "@types/koa__router": "^12.0.0",
```
Comment (antooni) [inline packages/backend/package.json:60]: is this reordering needed?

File: packages/backend/src/api/routers/TvlRouter.ts [hunk 9]
```diff
@@ -18,6 +18,43 @@ export function createTvlRouter(
 ) {
   const router = new Router()
 
+  // endpoint that will return tvl of specific projects
```
Comment (antooni) [inline packages/backend/src/api/routers/TvlRouter.ts:21]: remove this comment

File: packages/backend/src/api/controllers/tvl/DetailedTvlController.ts [hunk 10]
```diff
@@ -149,6 +145,69 @@ export class DetailedTvlController {
...
+
+    const tvlApiProjectResponse = generateDetailedAggregatedApiResponse(
+      groupedHourlyReports,
+      groupedSixHourlyReportsTree,
+      groupedDailyReports,
+      this.projects
+        .map((project) => project.projectId)
+        .filter(
+          (projectId) =>
+            projectIdsFilter.length === 0 ||
```
Comment (antooni) [inline packages/backend/src/api/controllers/tvl/DetailedTvlController.ts:200]: is this line added for optimization purposes?

File: packages/backend/src/api/controllers/tvl/generateTvlApiResponse.test.ts [hunk 11]
```diff
@@ -15,6 +15,106 @@ import { ReportRecord } from '../../../peripherals/database/ReportRepository'
...
+      })
+      result[projectId.toString()].push({
+        assetId,
+        tvl: asNumber(balanceUsd, 2),
+      })
+    }
+  }
+  return result
+}
 describe(generateTvlApiResponse.name, () => {
```
Comment (antooni) [inline packages/backend/src/api/controllers/tvl/generateTvlApiResponse.test.ts:118]: why was this function put at the end of the file?

File: packages/backend/src/api/controllers/tvl/types.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+interface SuccessResult<T> {
+  result: 'success'
```
Comment (torztomasz) [inline packages/backend/src/api/controllers/tvl/types.ts:2]: lets change 
```suggestion
  success: true
```

File: packages/backend/src/api/controllers/tvl/generateDetailedTvlApiResponse.ts [hunk 13]
```diff
@@ -64,7 +64,79 @@ export function generateDetailedTvlApiResponse(
...
+      first[3] + second[3],
+      first[4] + second[4],
+      first[5] + second[5],
+      first[6] + second[6],
+      first[7] + second[7],
+      first[8] + second[8],
+    ]
+  }
+
+  // for (const timestamp in hourlyGroupedByTimestamp) {
```
Comment (antooni) [inline packages/backend/src/api/controllers/tvl/generateDetailedTvlApiResponse.ts:103]: remove comment

File: packages/backend/src/api/routers/TvlRouter.ts [hunk 14]
```diff
@@ -18,6 +18,43 @@ export function createTvlRouter(
...
+  // accepts projectIds[] as query params
+  // /api/tvl/projects + projectIds[]
```
Comment (torztomasz) [inline packages/backend/src/api/routers/TvlRouter.ts:23]: /api/tvl/aggregate?projectSlugs=arbitrum,nova

**Review by torztomasz — CHANGES_REQUESTED** [review]

File: packages/backend/src/api/controllers/tvl/generateDetailedTvlApiResponse.ts [hunk 15]
```diff
@@ -64,7 +64,79 @@ export function generateDetailedTvlApiResponse(
...
+      {
+        hourly,
+        sixHourly,
+        daily,
+      },
+      projectId,
+    )
+  })
+
+  const firstProjectDetailedCharts = projectDetailedCharts[0]
```
Comment (antooni) [inline packages/backend/src/api/controllers/tvl/generateDetailedTvlApiResponse.ts:115]: is this code needed?

File: packages/backend/src/api/controllers/tvl/DetailedTvlController.ts [hunk 16]
```diff
@@ -149,6 +145,72 @@ export class DetailedTvlController {
     }
   }
 
+  async getDetailedAggregatedApiResponse(
+    slugs: string[],
+  ): Promise<TvlProjectResult> {
+    const dataTimings = await this.getDataTimings()
```
Comment (antooni) [inline packages/backend/src/api/controllers/tvl/DetailedTvlController.ts:151]: move this function above the if with condition : `!dataTimings.latestTimestamp`, this way we will not spend computations on getting the timings is the slugs are empty

File: packages/backend/src/api/controllers/tvl/generateDetailedTvlApiResponse.ts [hunk 17]
```diff
@@ -64,7 +65,75 @@ export function generateDetailedTvlApiResponse(
...
 }
 
-function getProjectCharts(
+export function generateDetailedAggregatedApiResponse(
+  hourly: ReportsPerProjectIdAndTimestamp,
+  sixHourly: ReportsPerProjectIdAndTimestamp,
+  daily: ReportsPerProjectIdAndTimestamp,
+  projectIds: ProjectId[],
+): DetailedTvlApiCharts {
+  const result: DetailedTvlApiCharts = {
```
Comment (antooni) [inline packages/backend/src/api/controllers/tvl/generateDetailedTvlApiResponse.ts:74]: extract this to const for readability

File: packages/backend/src/api/controllers/tvl/generateDetailedTvlApiResponse.ts [hunk 18]
```diff
@@ -64,7 +65,75 @@ export function generateDetailedTvlApiResponse(
...
+      data: [],
+    },
+    sixHourly: {
+      types: DETAILED_LABELS,
+      data: [],
+    },
+    daily: { types: DETAILED_LABELS, data: [] },
+  }
+
+  const mergeDetailValues = (
```
Comment (antooni) [inline packages/backend/src/api/controllers/tvl/generateDetailedTvlApiResponse.ts:86]: extract this to seperate function for readability

**Review by antooni — APPROVED** [review]

**Pull request merged**